### PR TITLE
Add Configuration Schema

### PR DESCRIPTION
**Pull request by mrocklin**
This adds type and description information to the configuration using the jsonschema spec.

So far this is only a proof of concept, and touches only a couple of entries.

Our plan is to use this short term for documentation, but in the future we might also validate user configuration at runtime.  

cc @jcrist @quasiben

### CODE REVIEW

**Comment by mrocklin** [comment]
Also, for documentation, there are projects that seem to marry sphinx with jsonschema

**Comment by mrocklin** [comment]
I tried using sphinx-jsonschema, but I'm not getting any output

**Comment by mrocklin** [comment]
OK, I tried out writing more of these.  It feels decently natural after a few minutes.

If we wanted to get clever we could start doing regex testing and such.  I'm going to leave this for future work though.

At this point I'm looking for approval from some other Dask maintainers that this is a direction that we want to go.  This is something that we might apply across many of the projects.  It would be good to have buy-in.

File: docs/source/configuration.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Configuration
+=============
+
+.. jsonschema:: https://raw.githubusercontent.com/jupyterhub/zero-to-jupyterhub-k8s/master/jupyterhub/schema.yaml
```
Comment (mrocklin) [inline docs/source/configuration.rst:4]: This is me just trying to get anything to print out.

File: distributed/distributed-schema.yaml [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        type: integer
+
+      scheduler:
+        type: object
+        properties:
+
+          allowed-failures:
+            type: integer
+            minimum: 0
+            description:
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/distributed-schema.yaml:16]: Ah, thanks.  I was wondering about this.  I remember now this from the helm days, but it's been a while and I definitely needed the reminder.

**Comment by mrocklin** [comment]
> I think this definitely makes sense for documentation for the core libraries (provided we can generate nice docs from it). If it's nice we might apply it elsewhere as well, but I don't think we should make it mandatory.

+1.  Fortunately it looks like jsonschema.validate passes on missing schema entries.

Any thoughts on a convention for where to place schema files?  Currently this PR does a `foo.yaml` -> `foo-schema.yaml`

**Comment by jcrist** [comment]
> Any thoughts on a convention for where to place schema files? Currently this PR does a foo.yaml -> foo-schema.yaml

If we're using it for testing and docs only, I might leave it out of the package. If you'd want to use it as part of the library (I'm not 100% sold on this yet, but could be) then adding a `schema` suffix or prefix makes sense.

**Comment by mrocklin** [comment]
> If we're using it for testing and docs only, I might leave it out of the package. If you'd want to use it as part of the library (I'm not 100% sold on this yet, but could be) then adding a schema suffix or prefix makes sense.

I'm mostly talking about a naming convention for the file.  My intent was for this to be an orthogonal consideration to "how much do we depend on this?".  Although my answer to the latter is "include it in the package, use it for docs, and later on figure out how to use it for run-time error detection."  I want to establish a convention on where to place this early on though so that we don't have to rename a bunch of files later on.

foo-schema.yaml ?

**Comment by jcrist** [comment]
> foo-schema.yaml

Makes sense to me.

**Comment by mrocklin** [comment]
OK, the process of describing config options here is complete.

File: docs/source/configuration.rst (hunk 1 above)
Comment (quasiben) [reply docs/source/configuration.rst:4]: I think we should remove the configuration page here in favor of https://github.com/dask/dask/pull/6069 then merge this in

File: docs/requirements.txt [hunk 3]
```diff
@@ -7,3 +7,4 @@ dask
 sphinx
 dask_sphinx_theme
 sphinx-click
+sphinx-jsonschema
```
Comment (quasiben) [inline docs/requirements.txt:10]: we can remove this if not using jsonschema

File: docs/source/configuration.rst (hunk 1 above)
Comment (mrocklin) [reply docs/source/configuration.rst]: Yup.  Thanks for flagging this.  I've removed all of the work around making an actual docpage and now this PR is just about the distributed-schema.yaml file, and testing its accuracy against the config file.  We don't do anything with documentation.

**Comment by quasiben** [comment]
Not sure why this is failing on windows.  This passes for me locally on OSX

**Comment by jrbourbeau** [comment]
It looks like the failure is related to the recently added `distributed.client.scheduler-info-interval` config option (see https://github.com/dask/distributed/pull/3723). We also added a `distributed.deploy.cluster-repr-interval` option in that PR

**Comment by mrocklin** [comment]
It is because the config was recently modified in a recent PR.  In this repository we're testing to ensure completeness of the schema.  I don't think that we necessarily want to do this in all of the repositories, but I think that it makes sense to do here.

I've modified this PR to now include an informative error message pointing future developers on how to resolve syncing issues when they arise.

```python-traceback
E           ValueError: 
E           The distributed.yaml and distributed-schema.yaml files are not in sync.
E           This usually happens when we add a new configuration value,
E           but don't add the schema of that value to the distributed-schema.yaml file
E           Please modify these files to include the missing values: 
E           
E               distributed.yaml:        ['cluster-repr-interval', 'lost-worker-timeout']
E               distributed-schema.yaml: ['lost-worker-timeout']
E           
E           Examples in these files should be a good start, 
E           even if you are not familiar with the jsonschema spec
```

**Comment by quasiben** [comment]
Thanks @mrocklin.  Merging in

**Pull request merged**